File: scripts/build_site.js
#!/usr/bin/env node

import { readFileSync } from "node:fs";
import { parseArgs as parseNodeArgs } from "node:util";
import { resolve } from "node:path";
import { atomicWrite, readCsv } from "./lib/csv.js";
import { finiteNumber, isMain, projectDirectory } from "./lib/runtime.js";

export const TIME_ZONE = "Europe/Ljubljana";
export const EXPECTED_COMPANIES = 10;
export const FONT_FILES = [
  "playfair-display-latin.woff2",
  "public-sans-latin.woff2",
  "OFL.txt",
];
export const REQUIRED_FIELDS = new Set([
  "rank", "ticker", "name", "sector", "subsector", "valuation_price", "price_date",
  "attractiveness_score", "quality_display_score", "quality_label", "why_selected",
  "valuation_summary", "business_quality_summary", "growth_summary", "balance_sheet_summary",
  "warning_summary", "what_to_verify", "review_status", "snapshot_source_url", "price_source_url",
]);

export function parseArgs(argv = process.argv.slice(2)) {
  const root = projectDirectory(import.meta.url);
  const { values } = parseNodeArgs({
    args: argv,
    options: {
      input: { type: "string", default: `${root}/generated/data/latest_research.csv` },
      templates: { type: "string", default: `${root}/templates` },
      assets: { type: "string", default: `${root}/assets` },
      output: { type: "string", default: `${root}/generated/site` },
      "published-at": { type: "string" },
      help: { type: "boolean", short: "h" },
    },
  });
  return {
    input: resolve(values.input), templates: resolve(values.templates), assets: resolve(values.assets),
    output: resolve(values.output), publishedAt: values["published-at"] ?? null,
    help: values.help ?? false,
  };
}

export function loadTemplate(path) {
  try {
    return readFileSync(path, "utf8");
  } catch (error) {
    throw new Error(`Cannot read template: ${path}`, { cause: error });
  }
}

export function number(row, field) {
  const value = finiteNumber(row[field]);
  if (value === null) throw new Error(`${row.ticker || "row"} has invalid ${field}.`);
  return value;
}

export function escapeHtml(value) {
  return String(value ?? "").trim()
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&#x27;");
}

export function safeUrl(row, field) {
  const value = String(row[field] ?? "").trim();
  let parsed;
  try {
    parsed = new URL(value);
  } catch {
    throw new Error(`${row.ticker || "row"} has invalid ${field}.`);
  }
  if (!["http:", "https:"].includes(parsed.protocol) || !parsed.hostname) {
    throw new Error(`${row.ticker || "row"} has invalid ${field}.`);
  }
  return escapeHtml(value);
}

export function loadRows(path) {
  let parsed;
  try {
    parsed = readCsv(path);
  } catch (error) {
    if (error.code === "ENOENT" || error.message.startsWith("Input file does not exist")) {
      throw new Error(`Input file does not exist: ${path}`);
    }
    throw error;
  }
  const missing = [...REQUIRED_FIELDS].filter((field) => !parsed.fields.includes(field)).sort();
  if (missing.length) throw new Error(`Input is missing required columns: ${missing.join(", ")}`);
  if (parsed.rows.length !== EXPECTED_COMPANIES) {
    throw new Error(`Expected ${EXPECTED_COMPANIES} research rows, found ${parsed.rows.length}.`);
  }

  const ranks = [];
  const tickers = new Set();
  for (const row of parsed.rows) {
    const rankText = String(row.rank ?? "").trim();
    const rank = Number(rankText);
    if (!/^[-+]?\d+$/.test(rankText) || !Number.isInteger(rank)) {
      throw new Error("Every research row must have an integer rank.");
    }
    const ticker = String(row.ticker ?? "").trim();
    if (!ticker || tickers.has(ticker)) throw new Error(`Ticker is empty or duplicated: ${ticker || "(empty)"}`);
    ranks.push(rank);
    tickers.add(ticker);
    number(row, "valuation_price");
    const attractiveness = number(row, "attractiveness_score");
    const quality = number(row, "quality_display_score");
    if (attractiveness < 0 || attractiveness > 100 || quality < 0 || quality > 10) {
      throw new Error(`${ticker} has a score outside its expected range.`);
    }
    safeUrl(row, "snapshot_source_url");
    safeUrl(row, "price_source_url");
  }
  const sortedRanks = [...ranks].sort((left, right) => left - right);
  if (sortedRanks.some((rank, index) => rank !== index + 1)) {
    throw new Error("Research ranks must be unique and consecutive from 1 to 10.");
  }
  return [...parsed.rows].sort((left, right) => Number(left.rank) - Number(right.rank));
}

function localParts(date) {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: TIME_ZONE,
    year: "numeric", month: "2-digit", day: "2-digit",
    hour: "2-digit", minute: "2-digit", second: "2-digit", hourCycle: "h23",
  }).formatToParts(date);
  return Object.fromEntries(parts.filter((part) => part.type !== "literal").map((part) => [part.type, part.value]));
}

function zoneOffsetMilliseconds(date, parts = localParts(date)) {
  return Date.UTC(
    Number(parts.year), Number(parts.month) - 1, Number(parts.day),
    Number(parts.hour), Number(parts.minute), Number(parts.second),
  ) - Math.floor(date.valueOf() / 1_000) * 1_000;
}

function offsetText(milliseconds) {
  const sign = milliseconds >= 0 ? "+" : "-";
  const totalMinutes = Math.abs(milliseconds) / 60_000;
  return `${sign}${String(Math.floor(totalMinutes / 60)).padStart(2, "0")}:${String(totalMinutes % 60).padStart(2, "0")}`;
}

export function publicationTime(value = null) {
  let moment;
  if (value) {
    if (!/(?:Z|[+-]\d{2}:\d{2})$/i.test(value)) {
      throw new Error("--published-at must include a timezone offset.");
    }
    moment = new Date(value);
    if (!Number.isFinite(moment.valueOf())) throw new Error("--published-at must be a valid ISO-8601 timestamp.");
  } else {
    moment = new Date();
  }
  const parts = localParts(moment);
  const offset = zoneOffsetMilliseconds(moment, parts);
  const localHourAsUtc = Date.UTC(
    Number(parts.year), Number(parts.month) - 1, Number(parts.day), Number(parts.hour), 0, 0,
  );
  const date = new Date(localHourAsUtc - offset);
  const truncatedParts = localParts(date);
  const truncatedOffset = zoneOffsetMilliseconds(date, truncatedParts);
  const iso = `${truncatedParts.year}-${truncatedParts.month}-${truncatedParts.day}T${truncatedParts.hour}:00:00${offsetText(truncatedOffset)}`;
  const dateDisplay = new Intl.DateTimeFormat("en-US", {
    timeZone: TIME_ZONE, weekday: "long", month: "long", day: "numeric", year: "numeric",
  }).format(date);
  const timeParts = new Intl.DateTimeFormat("en-US", {
    timeZone: TIME_ZONE, hour: "2-digit", hourCycle: "h23", timeZoneName: "short",
  }).formatToParts(date);
  const displayHour = timeParts.find((part) => part.type === "hour")?.value ?? truncatedParts.hour;
  // Some small-ICU Node builds render this as "GMT+2". Ljubljana has used
  // CET/CEST for the dates supported by this project, so keep output stable
  // across developer machines and GitHub Actions images.
  const zoneName = truncatedOffset === 7_200_000 ? "CEST"
    : truncatedOffset === 3_600_000 ? "CET"
      : timeParts.find((part) => part.type === "timeZoneName")?.value ?? "";
  return {
    date,
    iso,
    display: `${dateDisplay.replace(/, (?=\d{4}$)/, ", ")} · ${displayHour}:00 ${zoneName}`,
    year: truncatedParts.year,
  };
}

export function attractivenessLabel(score) {
  const display = score / 10;
  if (display >= 9.0) return "Excellent";
  if (display >= 8.5) return "Strong";
  if (display >= 8.0) return "Attractive";
  if (display >= 7.0) return "Good";
  return "Fair";
}

export function scoreTone(label) {
  if (["Excellent", "Strong"].includes(label)) return "strong";
  if (["Attractive", "Good"].includes(label)) return "good";
  if (label === "Fair") return "fair";
  return "weak";
}

export function tickerSlug(value) {
  const slug = value.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
  if (!slug) throw new Error(`Cannot create an HTML id from ticker: ${value}`);
  return slug;
}

export function templateValues(row) {
  const attractiveness = number(row, "attractiveness_score");
  const attrLabel = attractivenessLabel(attractiveness);
  const qualityLabel = String(row.quality_label ?? "").trim();
  return {
    rank: escapeHtml(row.rank), ticker: escapeHtml(row.ticker), ticker_slug: tickerSlug(row.ticker),
    name: escapeHtml(row.name), sector: escapeHtml(row.sector),
    subsector: escapeHtml(row.subsector || row.sector),
    price: `$${number(row, "valuation_price").toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`,
    attractiveness_display: (attractiveness / 10).toFixed(1), attractiveness_label: attrLabel,
    attractiveness_tone: scoreTone(attrLabel), quality_display: number(row, "quality_display_score").toFixed(1),
    quality_label: escapeHtml(qualityLabel), quality_tone: scoreTone(qualityLabel),
    why_selected: escapeHtml(row.why_selected), valuation_summary: escapeHtml(row.valuation_summary),
    business_quality_summary: escapeHtml(row.business_quality_summary), growth_summary: escapeHtml(row.growth_summary),
    balance_sheet_summary: escapeHtml(row.balance_sheet_summary), warning_summary: escapeHtml(row.warning_summary),
    what_to_verify: escapeHtml(row.what_to_verify),
    review_class: row.review_status.trim() === "priority_review" ? "priority-review" : "",
    snapshot_source_url: safeUrl(row, "snapshot_source_url"), price_source_url: safeUrl(row, "price_source_url"),
  };
}

export function substitute(template, values) {
  return template.replace(/\$\{([A-Za-z_][A-Za-z0-9_]*)\}/g, (match, key) => {
    if (!(key in values)) throw new Error(`Missing template value: ${key}`);
    return values[key];
  });
}

export function renderSite(rows, templatesDirectory, published, styles) {
  const page = loadTemplate(resolve(templatesDirectory, "index.html"));
  const desktopRow = loadTemplate(resolve(templatesDirectory, "partials/desktop_row.html"));
  const mobileCard = loadTemplate(resolve(templatesDirectory, "partials/mobile_card.html"));
  const analysis = loadTemplate(resolve(templatesDirectory, "partials/analysis.html"));
  const desktopParts = [];
  const mobileParts = [];
  for (const row of rows) {
    const values = templateValues(row);
    values.analysis_content = substitute(analysis, values);
    desktopParts.push(substitute(desktopRow, values).trimEnd());
    mobileParts.push(substitute(mobileCard, values).trimEnd());
  }
  return substitute(page, {
    company_count: String(rows.length), desktop_rows: desktopParts.join("\n"),
    mobile_cards: mobileParts.join("\n"), published_iso: published.iso,
    published_display: published.display, publication_year: published.year, styles,
  });
}

export function build(args) {
  const rows = loadRows(args.input);
  const published = publicationTime(args.publishedAt);
  let styles;
  let script;
  const fonts = new Map();
  try {
    styles = readFileSync(resolve(args.assets, "styles.css"), "utf8");
    script = readFileSync(resolve(args.assets, "site.js"), "utf8");
    for (const file of FONT_FILES) {
      fonts.set(file, readFileSync(resolve(args.assets, "fonts", file)));
    }
  } catch (error) {
    throw new Error(`Cannot read site assets from ${args.assets}`, { cause: error });
  }
  const rendered = renderSite(rows, args.templates, published, styles);
  atomicWrite(resolve(args.output, "index.html"), rendered);
  atomicWrite(resolve(args.output, "assets/site.js"), script);
  for (const [file, contents] of fonts) {
    atomicWrite(resolve(args.output, "assets/fonts", file), contents);
  }
  console.log(`Wrote ${resolve(args.output, "index.html")} with ${rows.length} companies (published ${published.iso}).`);
}

export async function main(argv = process.argv.slice(2)) {
  try {
    const args = parseArgs(argv);
    if (args.help) {
      console.log("Usage: node scripts/build_site.js [--input PATH] [--templates PATH] [--assets PATH] [--output PATH] [--published-at ISO]");
      return 0;
    }
    build(args);
    return 0;
  } catch (error) {
    console.error(`error: ${error.message}`);
    return 1;
  }
}

if (isMain(import.meta.url)) process.exitCode = await main();

File: scripts/build_site.py
#!/usr/bin/env python3

"""Build the static Value Stock Weekly site from latest_research.csv.

The generated ``generated/site`` directory can be published with GitHub Pages.
HTML lives in editable templates, while this script validates and formats the
data. Only the Python standard library is required.
"""

from __future__ import annotations

import argparse
import csv
import html
import math
import os
import re
import tempfile
from datetime import datetime
from pathlib import Path
from string import Template
from typing import Sequence
from urllib.parse import urlparse
from zoneinfo import ZoneInfo


TIME_ZONE = ZoneInfo("Europe/Ljubljana")
EXPECTED_COMPANIES = 10
REQUIRED_FIELDS = {
    "rank",
    "ticker",
    "name",
    "sector",
    "subsector",
    "valuation_price",
    "price_date",
    "attractiveness_score",
    "quality_display_score",
    "quality_label",
    "why_selected",
    "valuation_summary",
    "business_quality_summary",
    "growth_summary",
    "balance_sheet_summary",
    "warning_summary",
    "what_to_verify",
    "review_status",
    "snapshot_source_url",
    "price_source_url",
}


def parse_args(argv: Sequence[str] | None = None) -> argparse.Namespace:
    project_dir = Path(__file__).resolve().parent.parent
    parser = argparse.ArgumentParser(description="Build the static weekly shortlist site.")
    parser.add_argument(
        "--input",
        type=Path,
        default=project_dir / "generated" / "data" / "latest_research.csv",
        help="Research CSV (default: ./generated/data/latest_research.csv).",
    )
    parser.add_argument(
        "--templates",
        type=Path,
        default=project_dir / "templates",
        help="Template directory (default: ./templates).",
    )
    parser.add_argument(
        "--assets",
        type=Path,
        default=project_dir / "assets",
        help="Source asset directory (default: ./assets).",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=project_dir / "generated" / "site",
        help="GitHub Pages output directory (default: ./generated/site).",
    )
    parser.add_argument(
        "--published-at",
        help="Optional ISO-8601 timestamp for reproducible builds; defaults to now.",
    )
    return parser.parse_args(argv)


def load_template(path: Path) -> Template:
    try:
        return Template(path.read_text(encoding="utf-8"))
    except OSError as error:
        raise RuntimeError(f"Cannot read template: {path}") from error


def number(row: dict[str, str], field: str) -> float:
    try:
        value = float((row.get(field) or "").strip())
    except ValueError as error:
        raise RuntimeError(f"{row.get('ticker', 'row')} has invalid {field}.") from error
    if not math.isfinite(value):
        raise RuntimeError(f"{row.get('ticker', 'row')} has invalid {field}.")
    return value


def safe_url(row: dict[str, str], field: str) -> str:
    value = (row.get(field) or "").strip()
    parsed = urlparse(value)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise RuntimeError(f"{row.get('ticker', 'row')} has invalid {field}.")
    return html.escape(value, quote=True)


def load_rows(path: Path) -> list[dict[str, str]]:
    if not path.exists():
        raise RuntimeError(f"Input file does not exist: {path}")
    with path.open(newline="", encoding="utf-8-sig") as handle:
        reader = csv.DictReader(handle)
        rows = list(reader)
        missing = REQUIRED_FIELDS - set(reader.fieldnames or [])
    if missing:
        raise RuntimeError(f"Input is missing required columns: {', '.join(sorted(missing))}")
    if len(rows) != EXPECTED_COMPANIES:
        raise RuntimeError(
            f"Expected {EXPECTED_COMPANIES} research rows, found {len(rows)}."
        )

    ranks: list[int] = []
    tickers: set[str] = set()
    for row in rows:
        try:
            rank = int((row.get("rank") or "").strip())
        except ValueError as error:
            raise RuntimeError("Every research row must have an integer rank.") from error
        ticker = (row.get("ticker") or "").strip()
        if not ticker or ticker in tickers:
            raise RuntimeError(f"Ticker is empty or duplicated: {ticker or '(empty)'}")
        ranks.append(rank)
        tickers.add(ticker)
        number(row, "valuation_price")
        attractiveness = number(row, "attractiveness_score")
        quality = number(row, "quality_display_score")
        if not 0 <= attractiveness <= 100 or not 0 <= quality <= 10:
            raise RuntimeError(f"{ticker} has a score outside its expected range.")
        safe_url(row, "snapshot_source_url")
        safe_url(row, "price_source_url")
    if sorted(ranks) != list(range(1, EXPECTED_COMPANIES + 1)):
        raise RuntimeError("Research ranks must be unique and consecutive from 1 to 10.")
    return sorted(rows, key=lambda row: int(row["rank"]))


def publication_time(value: str | None) -> datetime:
    if value:
        normalized = value[:-1] + "+00:00" if value.endswith("Z") else value
        try:
            moment = datetime.fromisoformat(normalized)
        except ValueError as error:
            raise RuntimeError("--published-at must be a valid ISO-8601 timestamp.") from error
        if moment.tzinfo is None:
            raise RuntimeError("--published-at must include a timezone offset.")
        moment = moment.astimezone(TIME_ZONE)
    else:
        moment = datetime.now(TIME_ZONE)
    return moment.replace(minute=0, second=0, microsecond=0)


def attractiveness_label(score: float) -> str:
    display = score / 10
    if display >= 9.0:
        return "Excellent"
    if display >= 8.5:
        return "Strong"
    if display >= 8.0:
        return "Attractive"
    if display >= 7.0:
        return "Good"
    return "Fair"


def score_tone(label: str) -> str:
    if label in {"Excellent", "Strong"}:
        return "strong"
    if label in {"Attractive", "Good"}:
        return "good"
    if label == "Fair":
        return "fair"
    return "weak"


def escaped(value: str | None) -> str:
    return html.escape((value or "").strip())


def ticker_slug(value: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")
    if not slug:
        raise RuntimeError(f"Cannot create an HTML id from ticker: {value}")
    return slug


def template_values(row: dict[str, str]) -> dict[str, str]:
    attractiveness = number(row, "attractiveness_score")
    attr_label = attractiveness_label(attractiveness)
    quality_label = (row.get("quality_label") or "").strip()
    return {
        "rank": escaped(row["rank"]),
        "ticker": escaped(row["ticker"]),
        "ticker_slug": ticker_slug(row["ticker"]),
        "name": escaped(row["name"]),
        "sector": escaped(row["sector"]),
        "subsector": escaped(row["subsector"] or row["sector"]),
        "price": f"${number(row, 'valuation_price'):,.2f}",
        "attractiveness_display": f"{attractiveness / 10:.1f}",
        "attractiveness_label": attr_label,
        "attractiveness_tone": score_tone(attr_label),
        "quality_display": f"{number(row, 'quality_display_score'):.1f}",
        "quality_label": escaped(quality_label),
        "quality_tone": score_tone(quality_label),
        "why_selected": escaped(row["why_selected"]),
        "valuation_summary": escaped(row["valuation_summary"]),
        "business_quality_summary": escaped(row["business_quality_summary"]),
        "growth_summary": escaped(row["growth_summary"]),
        "balance_sheet_summary": escaped(row["balance_sheet_summary"]),
        "warning_summary": escaped(row["warning_summary"]),
        "what_to_verify": escaped(row["what_to_verify"]),
        "review_class": (
            "priority-review" if row["review_status"].strip() == "priority_review" else ""
        ),
        "snapshot_source_url": safe_url(row, "snapshot_source_url"),
        "price_source_url": safe_url(row, "price_source_url"),
    }


def render_site(
    rows: list[dict[str, str]], templates_dir: Path, published: datetime
) -> str:
    page = load_template(templates_dir / "index.html")
    desktop_row = load_template(templates_dir / "partials" / "desktop_row.html")
    mobile_card = load_template(templates_dir / "partials" / "mobile_card.html")
    analysis = load_template(templates_dir / "partials" / "analysis.html")

    desktop_parts: list[str] = []
    mobile_parts: list[str] = []
    for row in rows:
        values = template_values(row)
        analysis_html = analysis.substitute(values)
        values["analysis_content"] = analysis_html
        desktop_parts.append(desktop_row.substitute(values).rstrip())
        mobile_parts.append(mobile_card.substitute(values).rstrip())

    published_display = (
        f"{published.strftime('%A, %B')} {published.day}, {published.year} · "
        f"{published.strftime('%H:00 %Z')}"
    )
    return page.substitute(
        company_count=str(len(rows)),
        desktop_rows="\n".join(desktop_parts),
        mobile_cards="\n".join(mobile_parts),
        published_iso=published.isoformat(),
        published_display=published_display,
        publication_year=str(published.year),
    )


def atomic_write(path: Path, content: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary: Path | None = None
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", encoding="utf-8", dir=path.parent, delete=False
        ) as handle:
            handle.write(content)
            handle.flush()
            os.fsync(handle.fileno())
            temporary = Path(handle.name)
        os.replace(temporary, path)
    finally:
        if temporary and temporary.exists():
            temporary.unlink()


def build(args: argparse.Namespace) -> None:
    rows = load_rows(args.input)
    published = publication_time(args.published_at)
    rendered = render_site(rows, args.templates, published)

    style_source = args.assets / "styles.css"
    script_source = args.assets / "site.js"
    try:
        styles = style_source.read_text(encoding="utf-8")
        script = script_source.read_text(encoding="utf-8")
    except OSError as error:
        raise RuntimeError(f"Cannot read site assets from {args.assets}") from error

    atomic_write(args.output / "index.html", rendered)
    atomic_write(args.output / "assets" / "styles.css", styles)
    atomic_write(args.output / "assets" / "site.js", script)
    print(
        f"Wrote {args.output / 'index.html'} with {len(rows)} companies "
        f"(published {published.isoformat()})."
    )


def main(argv: Sequence[str] | None = None) -> int:
    try:
        build(parse_args(argv))
    except (OSError, RuntimeError, KeyError) as error:
        print(f"error: {error}", file=os.sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
